File: dart/crypto/lib/impl/kes_sum.dart
import 'dart:typed_data';

import 'package:blockchain_crypto/ed25519.dart';
import 'package:blockchain_crypto/impl/kes_helper.dart';
import 'package:blockchain_crypto/utils.dart';
import 'package:fixnum/fixnum.dart';
import 'package:fpdart/fpdart.dart';

/**
 * Credit to Aaron Schutza
 */
class KesSum {
  const KesSum();
  Future<KeyPairKesSum> createKeyPair(
      Uint8List seed, int height, Int64 offset) async {
    final tree = await generateSecretKey(seed, height);
    final vk = await generateVerificationKey(tree);
    return KeyPairKesSum(
        sk: SecretKeyKesSum(tree: tree, offset: offset), vk: vk);
  }

  Future<SignatureKesSum> sign(KesBinaryTree skTree, List<int> message) async {
    Future<SignatureKesSum> loop(
        KesBinaryTree keyTree, List<List<int>> W) async {
      if (keyTree is KesMerkleNode) {
        if (keyTree.left is KesEmpty)
          return loop(keyTree.right, [List.of(keyTree.witnessLeft)]..addAll(W));
        else
          return loop(keyTree.left, [List.of(keyTree.witnessRight)]..addAll(W));
      } else if (keyTree is KesSigningLeaf) {
        return SignatureKesSum(
            vk: Uint8List.fromList(keyTree.vk),
            signature:
                Uint8List.fromList(await ed25519.sign(message, keyTree.sk)),
            witness: W.map(Uint8List.fromList).toList());
      } else {
        return SignatureKesSum(
            vk: Uint8List(32),
            signature: Uint8List(64),
            witness: [Uint8List(0)]);
      }
    }

    return loop(skTree, []);
  }

  Future<bool> verify(SignatureKesSum signature, List<int> message,
      VerificationKeyKesSum vk) async {
    bool leftGoing(int level) => ((vk.step ~/ kesHelper.exp(level)) % 2) == 0;
    Future<bool> emptyWitness() async =>
        vk.value.sameElements(await kesHelper.hash(signature.vk));
    Future<bool> singleWitness(List<int> witness) async {
      final hashVkSign = await kesHelper.hash(signature.vk);
      if (leftGoing(0)) {
        return vk.value
            .sameElements(await kesHelper.hash(hashVkSign + witness));
      } else
        return vk.value
            .sameElements(await kesHelper.hash(witness + hashVkSign));
    }

    Future<bool> multiWitness(List<List<int>> witnessList,
        List<int> witnessLeft, List<int> witnessRight, int index) async {
      if (witnessList.isEmpty)
        return vk.value
            .sameElements(await kesHelper.hash(witnessLeft + witnessRight));
      else if (leftGoing(index))
        return multiWitness(
          witnessList.sublist(1),
          await kesHelper.hash(witnessLeft + witnessRight),
          witnessList.first,
          index + 1,
        );
      else
        return multiWitness(
          witnessList.sublist(1),
          witnessList.first,
          await kesHelper.hash(witnessLeft + witnessRight),
          index + 1,
        );
    }

    Future<bool> verifyMerkle(List<List<int>> W) async {
      if (W.isEmpty)
        return emptyWitness();
      else if (W.length == 1)
        return singleWitness(W.first);
      else if (leftGoing(0))
        return multiWitness(
            W.sublist(1), await kesHelper.hash(signature.vk), W.first, 1);
      else
        return multiWitness(
            W.sublist(1), W.first, await kesHelper.hash(signature.vk), 1);
    }

    final ed25519Verification =
        await ed25519.verify(signature.signature, message, signature.vk);
    if (!ed25519Verification) return false;

    final merkleVerification = await verifyMerkle(signature.witness);
    return merkleVerification;
  }

  Future<KesBinaryTree> update(KesBinaryTree tree, int step) async {
    if (step == 0) return tree;
    final totalSteps = kesHelper.exp(kesHelper.getTreeHeight(tree));
    final keyTime = getCurrentStep(tree);
    if (step < totalSteps && keyTime < step) {
      return await evolveKey(tree, step);
    }
    throw Exception(
        "Update error - Max steps: $totalSteps, current step: $keyTime, requested increase: $step");
  }

  int getCurrentStep(KesBinaryTree tree) {
    if (tree is KesMerkleNode) {
      if (tree.left is KesEmpty && tree.right is KesSigningLeaf)
        return 1;
      else if (tree.left is KesEmpty && tree.right is KesMerkleNode)
        return getCurrentStep(tree.right) +
            kesHelper.exp(kesHelper.getTreeHeight(tree.right));
      else if (tree.right is KesEmpty) return getCurrentStep(tree.left);
    }
    return 0;
  }

  Future<KesBinaryTree> generateSecretKey(Uint8List seed, int height) async {
    Future<KesBinaryTree> seedTree(Uint8List seed, int height) async {
      if (height == 0) {
        final keyPair = await ed25519.generateKeyPairFromSeed(seed);
        return KesSigningLeaf(
            Uint8List.fromList(keyPair.sk), Uint8List.fromList(keyPair.vk));
      } else {
        final r = await kesHelper.prng(seed);
        final left = await seedTree(r.first, height - 1);
        final right = await seedTree(r.second, height - 1);
        return KesMerkleNode(r.second, await kesHelper.witness(left),
            await kesHelper.witness(right), left, right);
      }
    }

    KesBinaryTree reduceTree(KesBinaryTree fullTree) {
      if (fullTree is KesMerkleNode) {
        eraseOldNode(fullTree.right);
        return KesMerkleNode(fullTree.seed, fullTree.witnessLeft,
            fullTree.witnessRight, reduceTree(fullTree.left), KesEmpty());
      } else {
        return fullTree;
      }
    }

    final out = reduceTree(await seedTree(seed, height));
    kesHelper.overwriteBytes(seed);
    return out;
  }

  Future<VerificationKeyKesSum> generateVerificationKey(
      KesBinaryTree tree) async {
    if (tree is KesMerkleNode) {
      return VerificationKeyKesSum(
          value: await kesHelper.witness(tree), step: getCurrentStep(tree));
    } else if (tree is KesSigningLeaf) {
      return VerificationKeyKesSum(
          value: await kesHelper.witness(tree), step: 0);
    } else {
      return VerificationKeyKesSum(value: Uint8List(32), step: 0);
    }
  }

  void eraseOldNode(KesBinaryTree node) {
    if (node is KesMerkleNode) {
      kesHelper.overwriteBytes(node.seed);
      kesHelper.overwriteBytes(node.witnessLeft);
      kesHelper.overwriteBytes(node.witnessRight);
      eraseOldNode(node.left);
      eraseOldNode(node.right);
    } else if (node is KesSigningLeaf) {
      kesHelper.overwriteBytes(node.sk);
      kesHelper.overwriteBytes(node.vk);
    }
  }

  Future<KesBinaryTree> evolveKey(KesBinaryTree input, int step) async {
    final halfTotalSteps = kesHelper.exp(kesHelper.getTreeHeight(input) - 1);
    shiftStep(int step) => step % halfTotalSteps;
    if (step >= halfTotalSteps) {
      if (input is KesMerkleNode) {
        if (input.left is KesSigningLeaf && input.right is KesEmpty) {
          final keyPair = await ed25519.generateKeyPairFromSeed(input.seed);
          final newNode = KesMerkleNode(
              Uint8List(input.seed.length),
              input.witnessLeft,
              input.witnessRight,
              KesEmpty(),
              KesSigningLeaf(Uint8List.fromList(keyPair.sk),
                  Uint8List.fromList(keyPair.vk)));
          eraseOldNode(input.left);
          kesHelper.overwriteBytes(input.seed);
          return newNode;
        }
        if (input.left is KesMerkleNode && input.right is KesEmpty) {
          final newNode = KesMerkleNode(
            Uint8List(input.seed.length),
            input.witnessLeft,
            input.witnessRight,
            KesEmpty(),
            await evolveKey(
                await generateSecretKey(
                    input.seed, kesHelper.getTreeHeight(input) - 1),
                shiftStep(step)),
          );
          eraseOldNode(input.left);
          kesHelper.overwriteBytes(input.seed);
          return newNode;
        }
        if (input.left is KesEmpty) {
          return KesMerkleNode(
              input.seed,
              input.witnessLeft,
              input.witnessRight,
              KesEmpty(),
              await evolveKey(input.right, shiftStep(step)));
        }
      } else if (input is KesSigningLeaf)
        return input;
      else
        return KesEmpty();
    } else {
      if (input is KesMerkleNode && input.right is KesEmpty) {
        return KesMerkleNode(input.seed, input.witnessLeft, input.witnessRight,
            await evolveKey(input.left, shiftStep(step)), KesEmpty());
      } else if (input is KesMerkleNode && input.left is KesEmpty) {
        return KesMerkleNode(input.seed, input.witnessLeft, input.witnessRight,
            KesEmpty(), await evolveKey(input.right, shiftStep(step)));
      } else if (input is KesSigningLeaf) {
        return input;
      }
      return KesEmpty();
    }
    return KesEmpty();
  }
}

const kesSum = KesSum();

abstract class KesBinaryTree {}

class KesMerkleNode extends KesBinaryTree {
  final Uint8List seed;
  final Uint8List witnessLeft;
  final Uint8List witnessRight;
  final KesBinaryTree left;
  final KesBinaryTree right;

  KesMerkleNode(
      this.seed, this.witnessLeft, this.witnessRight, this.left, this.right);

  @override
  int get hashCode => Object.hash(seed, witnessLeft, witnessRight, left, right);

  @override
  bool operator ==(Object other) {
    if (other is KesMerkleNode) {
      return seed.sameElements(other.seed) &&
          witnessLeft.sameElements(other.witnessLeft) &&
          witnessRight.sameElements(other.witnessRight) &&
          left == other.left &&
          right == other.right;
    }
    return false;
  }
}

class KesSigningLeaf extends KesBinaryTree {
  final Uint8List sk;
  final Uint8List vk;

  KesSigningLeaf(this.sk, this.vk);

  @override
  int get hashCode => Object.hash(sk, vk);

  @override
  bool operator ==(Object other) {
    if (other is KesSigningLeaf) {
      return sk.sameElements(other.sk) && vk.sameElements(other.vk);
    }
    return false;
  }
}

class KesEmpty extends KesBinaryTree {
  @override
  int get hashCode => 0;

  @override
  bool operator ==(Object other) {
    if (other is KesEmpty) {
      return true;
    }
    return false;
  }
}

class SecretKeyKesSum {
  final KesBinaryTree tree;
  final Int64 offset;

  SecretKeyKesSum({required this.tree, required this.offset});

  @override
  int get hashCode => Object.hash(tree, offset);

  @override
  bool operator ==(Object other) {
    if (other is SecretKeyKesSum) {
      return tree == other.tree && offset == other.offset;
    }
    return false;
  }
}

class VerificationKeyKesSum {
  final Uint8List value;
  final int step;

  VerificationKeyKesSum({required this.value, required this.step});
  @override
  int get hashCode => Object.hash(value, step);

  @override
  bool operator ==(Object other) {
    if (other is VerificationKeyKesSum) {
      return value.sameElements(other.value) && step == other.step;
    }
    return false;
  }
}

class KeyPairKesSum {
  final SecretKeyKesSum sk;
  final VerificationKeyKesSum vk;

  KeyPairKesSum({required this.sk, required this.vk});

  @override
  int get hashCode => Object.hash(sk, vk);

  @override
  bool operator ==(Object other) {
    if (other is KeyPairKesSum) {
      return sk == other.sk && vk == other.vk;
    }
    return false;
  }
}

class SignatureKesSum {
  final Uint8List vk;
  final Uint8List signature;
  final List<Uint8List> witness;

  SignatureKesSum({
    required this.vk,
    required this.signature,
    required this.witness,
  });

  @override
  int get hashCode => Object.hash(vk, signature, witness);

  @override
  bool operator ==(Object other) {
    if (other is SignatureKesSum) {
      return vk.sameElements(other.vk) &&
          signature.sameElements(other.signature) &&
          witness.length == other.witness.length &&
          witness.zip(other.witness).all((t) => t.first.sameElements(t.second));
    }
    return false;
  }
}

File: scala/core/src/main/scala/blockchain/crypto/KesSum.scala
package blockchain.crypto

import blockchain.crypto.KesBinaryTree._

import java.security.SecureRandom
import scala.annotation.tailrec

class KesSum extends SumComposition {

  def createKeyPair(
      seed: Array[Byte],
      height: Int,
      offset: Long
  ): (SecretKeyKesSum, VerificationKeyKesSum) =
    val sk: KesBinaryTree = generateSecretKey(seed, height)
    val pk: (Array[Byte], Int) = generateVerificationKey(sk)
    (SecretKeyKesSum(sk, offset), VerificationKeyKesSum(pk._1, pk._2))

  def sign(
      privateKey: SecretKeyKesSum,
      message: Array[Byte]
  ): SignatureKesSum =
    val sumSig = sign(privateKey.tree, message)
    SignatureKesSum(
      sumSig._1,
      sumSig._2,
      sumSig._3
    )

  def verify(
      signature: SignatureKesSum,
      message: Array[Byte],
      verifyKey: VerificationKeyKesSum
  ): Boolean =
    val sumSig =
      (
        signature.verificationKey,
        signature.signature,
        signature.witness.toVector
      )
    val sumVk = (verifyKey.value, verifyKey.step)
    verify(sumSig, message, sumVk)

  def update(privateKey: SecretKeyKesSum, steps: Int): SecretKeyKesSum =
    privateKey.copy(tree = updateKey(privateKey.tree, steps))

  def getCurrentStep(privateKay: SecretKeyKesSum): Int = getKeyTime(
    privateKay.tree
  )

  def getMaxStep(privateKay: SecretKeyKesSum): Int = exp(
    getTreeHeight(privateKay.tree)
  )

  def getVerificationKey(privateKey: SecretKeyKesSum): VerificationKeyKesSum =
    val vk = generateVerificationKey(privateKey.tree)
    VerificationKeyKesSum(vk._1, vk._2)

}

case class SecretKeyKesSum(tree: KesBinaryTree, offset: Long)

case class SignatureKesSum(
    verificationKey: Array[Byte],
    signature: Array[Byte],
    witness: Seq[Array[Byte]]
) {

  override def hashCode(): Int = {
    var r = 1
    r = 31 * r + java.util.Arrays.hashCode(verificationKey)
    r = 31 * r + java.util.Arrays.hashCode(signature)
    witness.foreach(w => r = 31 * r + java.util.Arrays.hashCode(w))
    r
  }

  override def equals(other: Any): Boolean = other match {
    case kesSum: SignatureKesSum =>
      verificationKey.sameElements(kesSum.verificationKey) &&
      signature.sameElements(kesSum.signature) &&
      witness.zip(kesSum.witness).forall { case (x, y) => x.sameElements(y) }

    case _ => false
  }
}

case class VerificationKeyKesSum(value: Array[Byte], step: Int) {

  override def hashCode(): Int = {
    var r = 1
    r = 31 * r + java.util.Arrays.hashCode(value) + step.hashCode
    r = 31 * r + step.hashCode
    r
  }

  override def equals(other: Any): Boolean = other match {
    case vk: VerificationKeyKesSum =>
      value.sameElements(vk.value) &&
      step == vk.step
    case _ => false
  }
}

/** AMS 2021: Implementation of the MMM construction: Malkin, T., Micciancio, D.
  * and Miner, S. (2002) ‘Efficient generic forward-secure signatures with an
  * unbounded number of time periods’, Advances in Cryptology Eurocrypt ’02,
  * LNCS 2332, Springer, pp.400–417.
  *
  * Provides forward secure signatures that cannot be reforged with a leaked
  * private key that has been updated.
  *
  * Number of time steps is determined by logl argument upon key generation,
  * theoretically unbounded for log(l)/log(2) = 7 in the asymmetric product
  * composition assuming integer time steps.
  *
  * Sum composition is based on underlying signing routine and the number of
  * time steps is configurable by specifying a tree height log(l)/log(2),
  * yielding l time steps.
  *
  * Credit to Aaron Schutza
  */
//noinspection ScalaStyle
class SumComposition extends KesEd25519Blake2b256 {

  override type SIG = (Array[Byte], Array[Byte], Vector[Array[Byte]])
  override type VK = (Array[Byte], Int)
  override type SK = KesBinaryTree

  private val random = new SecureRandom()

  /** Get the current time step of a sum composition key
    *
    * @param keyTree
    *   binary tree key
    * @return
    *   time step
    */
  private[crypto] def getKeyTime(keyTree: SK): Int =
    keyTree match {
      case MerkleNode(_, _, _, Empty(), _: SigningLeaf) => 1
      case MerkleNode(_, _, _, Empty(), right: MerkleNode) =>
        getKeyTime(right) + exp(getTreeHeight(right))
      case MerkleNode(_, _, _, left, Empty()) => getKeyTime(left)
      case _                                  => 0
    }

  /** Gets the public key in the sum composition
    *
    * @param keyTree
    *   binary tree for which the key is to be calculated
    * @return
    *   binary array public key
    */
  private[crypto] def generateVerificationKey(keyTree: SK): VK = {
    val h = keyTree match {
      case node: MerkleNode  => (witness(node), getKeyTime(keyTree))
      case leaf: SigningLeaf => (witness(leaf), 0)
      case Empty()           => (Array.fill(hashBytes)(0: Byte), 0)
    }
    //    println(s"---------------------------start sum verification key ----------------")
    //    println(s"verification witness: ${Base58.encode(h._1)}, keyTime: ${h._2}")
    //    println(s"---------------------------end sum verification key ----------------")
    h
  }

  /** Generates keys in the sum composition, recursive functions construct the
    * tree in steps and the output is the leftmost branch
    *
    * @param seed
    *   input entropy for binary tree and keypair generation
    * @param i
    *   height of tree
    * @return
    *   binary tree at time step 0
    */
  private[crypto] def generateSecretKey(seed: Array[Byte], height: Int): SK = {

    // generate the binary tree with the pseudorandom number generator
    def seedTree(seed: Array[Byte], height: Int): KesBinaryTree =
      if (height == 0) {
        val (sk, vk) = sGenKeypair(seed)
        SigningLeaf(vk, sk)
      } else {
        val r = prng(seed)
        val left = seedTree(r._1, height - 1)
        val right = seedTree(r._2, height - 1)
        MerkleNode(r._2, witness(left), witness(right), left, right)
      }

    // traverse down the tree to the leftmost leaf
    def reduceTree(fullTree: KesBinaryTree): KesBinaryTree =
      fullTree match {
        case MerkleNode(seed, witL, witR, nodeL, nodeR) =>
          eraseOldNode(nodeR)
          MerkleNode(seed, witL, witR, reduceTree(nodeL), Empty())
        case leaf: SigningLeaf => leaf
        case _                 => Empty()
      }

    // executes the above functions in order
    val out = reduceTree(seedTree(seed, height))
    random.nextBytes(seed)
    out
  }

  /** Updates the key in the sum composition
    *
    * @param keyTree
    *   binary tree to be updated
    * @param step
    *   time step key is to be updated to
    * @return
    *   updated key configuration
    */
  private[crypto] def updateKey(keyTree: SK, step: Int): SK = {
    val totalSteps = exp(getTreeHeight(keyTree))
    val keyTime = getKeyTime(keyTree)
    if (step == 0) keyTree
    else if (step < totalSteps && keyTime < step) {
      evolveKey(keyTree, step)
    } else {
      throw new Error(
        s"Update error - Max steps: $totalSteps, current step: $keyTime, requested increase: $step"
      )
    }
  }

  private[crypto] def eraseOldNode(node: KesBinaryTree): Unit =
    node match {
      case merkleNode: MerkleNode =>
        random.nextBytes(merkleNode.seed)
        random.nextBytes(merkleNode.witnessLeft)
        random.nextBytes(merkleNode.witnessRight)
        merkleNode.left match {
          case l: MerkleNode => eraseOldNode(l)
          case l: SigningLeaf =>
            random.nextBytes(l.sk)
            random.nextBytes(l.vk)
          case _ =>
        }
        merkleNode.right match {
          case r: MerkleNode => eraseOldNode(r)
          case r: SigningLeaf =>
            random.nextBytes(r.sk)
            random.nextBytes(r.vk)
          case _ =>
        }
      case leaf: SigningLeaf =>
        random.nextBytes(leaf.sk)
        random.nextBytes(leaf.vk)
      case _ =>
    }

  /** Evolves key a specified number of steps
    */
  private[crypto] def evolveKey(
      input: KesBinaryTree,
      step: Int
  ): KesBinaryTree = {
    val halfTotalSteps = exp(getTreeHeight(input) - 1)
    val shiftStep: Int => Int = (step: Int) => step % halfTotalSteps

    if (step >= halfTotalSteps) {
      input match {
        case MerkleNode(seed, witL, witR, oldLeaf: SigningLeaf, Empty()) =>
          val (sk, vk) = sGenKeypair(seed)
          val newNode =
            MerkleNode(
              Array.fill(seed.length)(0: Byte),
              witL,
              witR,
              Empty(),
              SigningLeaf(sk, vk)
            )
          eraseOldNode(oldLeaf)
          random.nextBytes(seed)
          newNode
        case MerkleNode(seed, witL, witR, oldNode: MerkleNode, Empty()) =>
          val newNode = MerkleNode(
            Array.fill(seed.length)(0: Byte),
            witL,
            witR,
            Empty(),
            evolveKey(
              generateSecretKey(seed, getTreeHeight(input) - 1),
              shiftStep(step)
            )
          )
          eraseOldNode(oldNode)
          random.nextBytes(seed)
          newNode
        case MerkleNode(seed, witL, witR, Empty(), right) =>
          MerkleNode(
            seed,
            witL,
            witR,
            Empty(),
            evolveKey(right, shiftStep(step))
          )

        case leaf: SigningLeaf => leaf
        case _                 => Empty()
      }
    } else {
      input match {
        case MerkleNode(seed, witL, witR, left, Empty()) =>
          MerkleNode(
            seed,
            witL,
            witR,
            evolveKey(left, shiftStep(step)),
            Empty()
          )

        case MerkleNode(seed, witL, witR, Empty(), right) =>
          MerkleNode(
            seed,
            witL,
            witR,
            Empty(),
            evolveKey(right, shiftStep(step))
          )

        case leaf: SigningLeaf => leaf
        case _                 => Empty()
      }
    }
  }

  /** Signature in the sum composition
    *
    * @param keyTree
    *   secret key tree of the sum composition
    * @param m
    *   message to be signed
    * @return
    *   byte array signature
    */
  private[crypto] def sign(keyTree: SK, m: Array[Byte]): SIG = {
    // loop that generates the signature of m and stacks up the witness path of the key
    @tailrec
    def loop(
        keyTree: KesBinaryTree,
        W: Vector[Array[Byte]] = Vector()
    ): SIG = keyTree match {
      case MerkleNode(_, witL, _, Empty(), right) =>
        loop(right, witL.clone() +: W)
      case MerkleNode(_, _, witR, left, _) => loop(left, witR.clone() +: W)
      case leaf: SigningLeaf => (leaf.vk.clone(), sSign(m, leaf.sk).clone(), W)
      case _ =>
        (
          Array.fill(pkBytes)(0: Byte),
          Array.fill(sigBytes)(0: Byte),
          Vector(Array())
        )
    }
    loop(keyTree)
  }

  /** Verify in the sum composition
    * @param kesSig
    *   signature to be verified
    * @param m
    *   message corresponding to the signature
    * @param kesVk
    *   verification key of the sum composition
    * @return
    *   true if the signature is valid false if otherwise
    */
  private[crypto] def verify(
      kesSig: SIG,
      m: Array[Byte],
      kesVk: VK
  ): Boolean = {
    val (vkSign, sigSign, merkleProof) = kesSig
    val (root: Array[Byte], step: Int) = kesVk

    // determine if the step corresponds to a right or left decision at each height
    val leftGoing: Int => Boolean = (level: Int) =>
      ((step / exp(level)) % 2) == 0

    def verifyMerkle(W: Vector[Array[Byte]]): Boolean =
      if (W.isEmpty) emptyWitness
      else if (W.length == 1) singleWitness(W.head)
      else if (leftGoing(0)) multiWitness(W.tail, hash(vkSign), W.head, 1)
      else multiWitness(W.tail, W.head, hash(vkSign), 1)

    def emptyWitness: Boolean = root sameElements hash(vkSign)

    def singleWitness(witness: Array[Byte]): Boolean =
      if (leftGoing(0)) root sameElements hash(hash(vkSign) ++ witness)
      else root sameElements hash(witness ++ hash(vkSign))

    @tailrec
    def multiWitness(
        witnessList: Vector[Array[Byte]],
        witnessLeft: Array[Byte],
        witnessRight: Array[Byte],
        index: Int
    ): Boolean =
      if (witnessList.isEmpty)
        root sameElements hash(witnessLeft ++ witnessRight)
      else if (leftGoing(index))
        multiWitness(
          witnessList.tail,
          hash(witnessLeft ++ witnessRight),
          witnessList.head,
          index + 1
        )
      else
        multiWitness(
          witnessList.tail,
          witnessList.head,
          hash(witnessLeft ++ witnessRight),
          index + 1
        )

    val verifySign = sVerify(m, sigSign, vkSign)

    verifyMerkle(merkleProof) && verifySign

  }
}
